packages/pigeon/lib/cpp_generator.dart

```diff
@@ -549,16 +785,51 @@ String? _cppTypeForBuiltinDartType(TypeDeclaration type) {
   }
 }
 
-String _cppTypeForDartType(TypeDeclaration type) {
-  return _cppTypeForBuiltinDartType(type) ?? type.baseName;
+/// Returns the base C++ type (without pointer, reference, optional, etc.) for
+/// the given [type].
+String _baseCppTypeForDartType(TypeDeclaration type) {
+  return _baseCppTypeForBuiltinDartType(type) ?? type.baseName;
+}
+
+/// Returns the C++ type to use in a value context (variable declaration,
+/// pass-by-value, etc.) for the given C++ base type.
+String _valueType(HostDatatype type) {
+  final String baseType = type.datatype;
+  return type.isNullable ? 'std::optional<$baseType>' : baseType;
+}
+
+/// Returns the C++ type to use in an argument context without ownership
+/// transfer for the given base type.
+String _unownedArgumentType(HostDatatype type) {
+  final bool isString = type.datatype == 'std::string';
+  final String baseType = isString ? 'std::string_view' : type.datatype;
+  if (isString || _isPodType(type)) {
+    return type.isNullable ? 'const $baseType*' : baseType;
+  }
+  return type.isNullable ? 'const $baseType*' : 'const $baseType&';
+}
+
+/// Returns the C++ type to use for the return of a getter for a field of type
+/// [type].
+String _getterReturnType(HostDatatype type) {
+  final String baseType = type.datatype;
+  if (_isPodType(type)) {
+    // Use pointers rather than optionals even for nullable POD, since the
+    // semantics of using them is essentially identical and this makes them
+    // consistent with non-POD.
+    return type.isNullable ? 'const $baseType*' : baseType;
+  }
+  return type.isNullable ? 'const $baseType*' : 'const $baseType&';
 }
 
+// TODO(stuartmorgan): Audit all uses of this and convert them to context-based
+// methods like those above. Code still using this method may well have bugs.
```

Comment:
I suspect that there are at least some bugs similar to the ones I fixed for field types in other places, but I wanted to limit the scope of this PR. We should get more of the native tests online in follow-ups to flush out other bugs.

Reply:
Yes. This PR removes fields from being unique_ptrs, so the use of this method should be greatly reduced here.